### ISSUE TITLE
RDataFrame interface to input data to RootInteracive as a replacement of tree2Panda (& uproot)

### ISSUE THREAD

**Issue by miranov25** [miranov25/RootInteractive#243]
The functionality to the tree- >Draw queries (used internally in tree2Panda) and RDataFrame does not overlap. To be able to use complex C++ structures in O2, we have to switch to RDataFrame.

New functionality within the scikit-hep - https://github.com/scikit-hep/awkward/pull/1295

### tree->Draw
- (+) easy to query
- (+) aliase for lazy evaluation
-- (+) dependency trees created automaticaly

-- (-) caching not supported
-- (-) only simple variables supported 
-- (-) slower as an C interpreter used

## RDataframe
- (+) fast -JIT functions supported
- (+) support for arbitrary C++ structures
- (-) complex aliases not supported
- (-) dependency trees to be provide by users

**Comment by ianna** [comment]
@miranov25 - FYI, Awkward <-> RdataFrame has been completed and fully integrated to Awkward master branch. Please, see a tutorial draft [here](https://github.com/ianna/PyHep2022/blob/main/Awkward%20to%20RDataFrame%20and%20back%20v1.ipynb).
Please, let me know if anything is missing or does not work. Thanks!

**Comment by miranov25** [comment]
Dear @ianna 

Thank you for the link. I went through the tutorial briefly. I have an idea for further steps. We could simplify the creation of a reusable library with the functions, similar to what we did in the past with the declaration of TTree::SetAlias. Perhaps this is already possible in ROOT v6.26. If this works as we envisioned, we could offer our RootInteractive visualisation and interactive ND histogramming/aggregation package (+ ML) for general use.

Before offering it within scikit-hep, we need to improve the documentation and clearly state whether it is an experimental (ML) or stable part of our package. We have considered PyHep 2022, but we are too late https://indico.cern.ch/e/PyHEP2022. We can use the time now for full, roundabout integration of RDataFrame < -- > into our RootInteractive tool, including the possibilities of client queries via joins.

If you are interested, I would like to ask for your advice and help. To get an idea of what the project is about, I would like to refer you to the RootIntreactive tutorial (March 2022):
https://indico.cern.ch/event/1135398/

There are quite a few use cases, so I would like to refer to a more detailed example presentation/Jupyter Notebook/Dashboard/accompanying video:
* Interactive dashboard: 
  *  https://indico.cern.ch/event/1135398/contributions/4764024/subcontributions/370740/attachments/2402507/4114266/CMITSimulationsGEMTPC.html
* Presentation:
  * https://indico.cern.ch/event/1135398/contributions/4764014/attachments/2405453/4114991/PWGPP-485NDPipelineRootInteractive_Tutorial10032022.pdf
  * https://indico.cern.ch/event/1135398/contributions/4764024/subcontributions/370740/attachments/2402507/4114272/CMITSimulGEMTPC_RootInteractiveTutorial10032022.pdf
* Demo film - RooInteractive part  at 16:00 
  * https://indico.cern.ch/event/1135398/contributions/4764024/subcontributions/370740/attachments/2402507/4109039/CMITSimulationsGEMTPC.mp4

The tutorial is 6 months old in meantime we improved the speed with the aim of interactive detector physics and physics analysis:

* Dashboard :
  *        https://indico.cern.ch/event/1135398/contributions/4950038/attachments/2474468/4245987/test_EffTrack.html* 
  
* Notebook:
  *        https://indico.cern.ch/event/1135398/#preview:4265612

I plan to visit CERN in September. In case of interest, we can meet before on zoom or at CERN. 

Marian

**Comment by ianna** [comment]
Hi Marian,

Let’s meet at CERN. Please, contact me by e-mail iannadotosborneatcerndotch

Kind regards,

Ianna

On 2 Sep 2022, at 15:13, miranov25 ***@***.******@***.***>> wrote:



Dear @ianna<https://github.com/ianna>

Thank you for the link. I went through the tutorial briefly. I have an idea for further steps. We could simplify the creation of a reusable library with the functions, similar to what we did in the past with the declaration of TTree::SetAlias. Perhaps this is already possible in ROOT v6.26. If this works as we envisioned, we could offer our RootInteractive visualisation and interactive ND histogramming/aggregation package (+ ML) for general use.

Before offering it within scikit-hep, we need to improve the documentation and clearly state whether it is an experimental (ML) or stable part of our package. We have considered PyHep 2022, but we are too late https://indico.cern.ch/e/PyHEP2022. We can use the time now for full, roundabout integration of RDataFrame < -- > into our RootInteractive tool, including the possibilities of client queries via joins.

If you are interested, I would like to ask for your advice and help. To get an idea of what the project is about, I would like to refer you to the RootIntreactive tutorial (March 2022):
https://indico.cern.ch/event/1135398/

There are quite a few use cases, so I would like to refer to a more detailed example presentation/Jupyter Notebook/Dashboard/accompanying video:

  *   Interactive dashboard:
     *   https://indico.cern.ch/event/1135398/contributions/4764024/subcontributions/370740/attachments/2402507/4114266/CMITSimulationsGEMTPC.html
  *   Presentation:
     *   https://indico.cern.ch/event/1135398/contributions/4764014/attachments/2405453/4114991/PWGPP-485NDPipelineRootInteractive_Tutorial10032022.pdf
     *   https://indico.cern.ch/event/1135398/contributions/4764024/subcontributions/370740/attachments/2402507/4114272/CMITSimulGEMTPC_RootInteractiveTutorial10032022.pdf
  *   Demo film - RooInteractive part at 16:00
     *   https://indico.cern.ch/event/1135398/contributions/4764024/subcontributions/370740/attachments/2402507/4109039/CMITSimulationsGEMTPC.mp4

The tutorial is 6 months old in meantime we improved the speed with the aim of interactive detector physics and physics analysis:

  *   Dashboard :

     *

   https://indico.cern.ch/event/1135398/contributions/4950038/attachments/2474468/4245987/test_EffTrack.html*


  *   Notebook:

     *

   https://indico.cern.ch/event/1135398/#preview:4265612


I plan to visit CERN in September. In case of interest, we can meet before on zoom or at CERN.

Marian

**Comment by miranov25** [comment]
###  Related JIRA ticket testing using "ToyMC" dEdx algorithm scan 

tested in RI Notebook using :
 * [https://gitlab.cern.ch/alice-tpc-offline/alice-tpc-notes/-/blob/ab61ebe1148ebc7bd88c03f667aff0caa3b2b03e/JIRA/ATO-614/code/toydEdxSimul.C] 
 * [https://gitlab.cern.ch/alice-tpc-offline/alice-tpc-notes/-/blob/611d5048fced221189c232cdc99eb953ab6bf470/JIRA/ATO-614/RDFtoAwkward.ipynb] 

Non-trivial RDataframe used for the dEdx algorithm optimization - translated to the RDataFrame <-> awkward
 
 * Check if the dump working 
   * OK
 * Check if it works parallel
   *  ROOT::EnableImplicitMT(32) - ???
   * Working very well - in example 32 parallel jobs 
```python
CPU times: user 1min 44s, sys: 884 ms, total: 1min 45s
Wall time: 10.2 s
```
 * Check if we can do ML training (RadnomForest)
 * Check if we can add the results of ML prediction back to the DataFrame and export to the tree

**Comment by miranov25** [comment]
## RDataframe <-> awkae questions:

Related Jupyter and dashboard:
https://gitlab.cern.ch/alice-tpc-offline/alice-tpc-notes/-/blob/1b6fd40f1620b6871a68d7b659c70e009eb0c028/JIRA/ATO-614/drawRDFGasGain.ipynb

### Open questions:

- Optimization of the Machine learning training. Awkward structure non-flat. Flattening converting to panda (with entry and subentry), to be able to use it in the ML training resp. prediction
  * Example use case for Machine learning using awkward 
  *  awkward in -> ML -> awkward out-> RDtataFrame- define
- Event loop and eventual disk usage
  * Not clear at all if possible 
  * RDataFrame - processing per "event"
  * Predictions to be done per batch ...
- If not possible - what are other options?

**Comment by miranov25** [comment]
Answers to questions:
* https://awkward-array.org/quickstart.html

**Comment by ianna** [comment]
> Answers to questions:
> 
> * https://awkward-array.org/quickstart.html
> * padding  reference (reference to )
>   
>   * https://github.com/scikit-hep/awkward

Please, try:
```python
>>> import awkward as ak
>>> from sklearn.ensemble import RandomForestClassifier
>>> clf = RandomForestClassifier(random_state=0)
>>> xarr = ak.Array([[1,2,3], [11,22,33]])
>>> yarr = ak.Array([0,1])
>>> clf.fit(xarr, yarr)
RandomForestClassifier(random_state=0)
>>> clf.predict(xarr)
array([0, 1])
>>> clf.predict([[4, 5, 6], [14, 15, 16]])
array([0, 1])
>>> 

```

**Comment by ianna** [comment]
> Answers to questions:
> 
> * https://awkward-array.org/quickstart.html
> * padding  reference (reference to )
>   
>   * https://github.com/scikit-hep/awkward

https://awkward-array.readthedocs.io/en/latest/_auto/ak.pad_none.html

**Comment by miranov25** [comment]
Link to tutorials:
https://github.com/ianna/PyHep2022

**Comment by miranov25** [comment]
## Link to chat in scikit-hep - discussion about limitations and possible improvement - array of structures

* current limitation of imports for the array of structures  (Ianna, Marian)
  * https://gitter.im/Scikit-HEP/awkward-array?at=6388d4714fee72663c9ed749
  * not supported for a moment
* Answer from Jim - consideration - what could be  supported
 * https://gitter.im/Scikit-HEP/awkward-array?at=638a1353ff5546644b702d3b
 * array of structures
    * RDataFrame -> awkward - not yet
    * awkward->RDataFrame - should be possible

**Comment by miranov25** [comment]
### RDataFrame column filter:

* https://github.com/miranov25/RootInteractive/blob/1f55494019e3a645c7d204d5b24e50f4883ff8bc/RootInteractive/Tools/RDataFrameTools.py 

```python
In [86]: filterRDFColumns?
Signature:
filterRDFColumns(
    rdf,
    selectList=['.*'],
    excludeList=[],
    selectTypeList=['.*'],
    excludeTypeList=['.*AliExternal.*'],
    verbose=0,
)
Docstring:
function to filter available columns in RDataFrame
:param rdf:               - input RDataFrame
:param selectList:        - columns to select (regExp)
:param excludeList:       - columns to reject (regExp)
:param selectTypeList     - types to accept   (regExp)
:param excludeTypeList:   - types to reject   (regExp)
:param verbose:           - verbosity 0x1 -print all status  0x2 - print selected  , 0x4 print rejected
:return:                    filtered list of columns

example:
filterRDFColumns(rdf1, ["param.*","delta","covar"],["part.",".*Refit.*"],[".*"],[""], verbose=1)
```

**Comment by miranov25** [comment]
## Discussion in gitter chat :
*https://gitter.im/matrix/5ba1f93bd73408ce4fa8a265/@agoose77:matrix.org?at=639e235d967c8305843b106c

## List of current supported features:
* https://gitter.im/matrix/5ba1f93bd73408ce4fa8a265/@agoose77:matrix.org?at=639e2f21a151003b5a72fb22

**Comment by miranov25** [comment]
## Discussion in gitter -simplifying RDataFrame:

https://gitter.im/matrix/5ba1f93bd73408ce4fa8a265/@agoose77:matrix.org?at=639f83faa151003b5a7550f4
